File: src/main/java/me/chriss99/worldmanagement/ChunkRegionFileManager.java
package me.chriss99.worldmanagement;

import me.chriss99.Array2DBufferWrapper;
import org.joml.Vector2i;

import java.nio.ByteBuffer;
import java.util.Map;

public class ChunkRegionFileManager implements RegionFileManager<Chunk> {
    private final FileLoadStoreManager<Region<Chunk>> fileManager;
    private final int chunkByteSize;
    private final int chunkDataByteSize;
    public final Array2DBufferWrapper.Type type;
    public final int chunkSize;

    public ChunkRegionFileManager(String worldName, Array2DBufferWrapper.Type type, int chunkSize) {
        this.fileManager = new FileLoadStoreManager<>("worlds/" + worldName, "region", this::regionFromByteArray, this::regionToByteArray);

        chunkDataByteSize = chunkSize*chunkSize*type.elementSize;
        chunkByteSize = chunkDataByteSize + 4*2;
        this.type = type;
        this.chunkSize = chunkSize;
    }

    public Region<Chunk> loadRegion(Vector2i regionCoord) {
        return fileManager.loadFile(regionCoord);
    }

    public void saveRegion(Region<Chunk> region) {
        fileManager.saveFile(region, region.coord);
    }

    private Region<Chunk> regionFromByteArray(byte[] array, Vector2i regionCoord) {
        int chunkNum = array.length/chunkByteSize;
        Region<Chunk> region = new Region<>(regionCoord);
        ByteBuffer buffer = ByteBuffer.wrap(array);

        for (int i = 0; i < chunkNum; i++) {
            Vector2i chunkCoord = new Vector2i(buffer.getInt(), buffer.getInt());
            byte[] byteArray = new byte[chunkDataByteSize];
            buffer.get(byteArray);
            Array2DBufferWrapper data = Array2DBufferWrapper.of(ByteBuffer.wrap(byteArray), type, chunkSize, chunkSize);
            region.addChunk(chunkCoord, new Chunk(data));
        }

        return region;
    }

    private byte[] regionToByteArray(Region<Chunk> region) {
        ByteBuffer buffer = ByteBuffer.allocate(region.getAllChunks().size()*chunkByteSize);

        for (Map.Entry<Vector2i, Chunk> entry : region.getAllChunks()) {
            buffer.putInt(entry.getKey().x);
            buffer.putInt(entry.getKey().y);
            buffer.put(entry.getValue().data().buffer);
        }

        return buffer.array();
    }
}

File: src/main/kotlin/worldmanagement/ChunkRegionFileManager.kt
package me.chriss99.worldmanagement

import me.chriss99.Array2DBufferWrapper
import glm_.vec2.Vec2i
import java.nio.ByteBuffer

class ChunkRegionFileManager(worldName: String, val type: Array2DBufferWrapper.Type, val chunkSize: Int) :
    AbstractRegionFileManager<Chunk>(worldName, "region") {
    private val chunkDataByteSize: Int = chunkSize * chunkSize * type.elementSize
    private val chunkByteSize: Int = chunkDataByteSize + 4 * 2

    override fun regionFromBytes(array: ByteArray, regionCoord: Vec2i): Region<Chunk> {
        val chunkNum = array.size / chunkByteSize
        val region: Region<Chunk> = Region(regionCoord)
        val buffer = ByteBuffer.wrap(array)

        for (i in 0..<chunkNum) {
            val chunkCoord: Vec2i = Vec2i(buffer.getInt(), buffer.getInt())
            val byteArray = ByteArray(chunkDataByteSize)
            buffer.get(byteArray)
            val data = Array2DBufferWrapper.of(ByteBuffer.wrap(byteArray), type, Vec2i(chunkSize))
            region.addChunk(chunkCoord, Chunk(data))
        }

        return region
    }

    override fun regionToBytes(region: Region<Chunk>): ByteArray {
        val buffer = ByteBuffer.allocate(region.allTiles.size * chunkByteSize)

        for ((key, value) in region.allTiles) {
            buffer.putInt(key.x)
            buffer.putInt(key.y)
            buffer.put(value.data.buffer)
        }

        return buffer.array()
    }
}